### ISSUE TITLE
Google Cloud Life Sciences Deprecation

### ISSUE THREAD

**Issue by cademirch** [snakemake/snakemake#2360]
Starting this to bring awareness to Google [deprecating the Cloud Life Sciences API](https://cloud.google.com/batch/docs/migrate-to-batch-from-cloud-life-sciences). As well as discuss migrating to Batch.

From Google:

> On July 17, 2023, Google Cloud announced that Cloud Life Sciences, which had been in beta, is deprecated. New projects cannot onboard to Cloud Life Sciences, and the service will no longer be available on Google Cloud after July 8, 2025. However, Batch is generally available and is a comprehensive successor that supports all use cases for Cloud Life Sciences.

As of today, if a user had not already had the API enabled, they will not be able to. So migrating to Batch should probably be a  priority. 

I'll try to carve out some time to look at this more closely, but would appreciate any thoughts and help.

**Comment by cademirch** [comment]
Thinking about this a little more today, it probably makes the most sense to build a new executor (i.e. `GoogleBatch`) and continue supporting the Life Sciences executor until its fully deprecated in 2025. Obviously building a new executor will take quite a bit of time and effort, but will allow us possibly to take advantage of some of Batch's features. I can probably devote a decent amount of time this summer, but would like to discuss and come up with something of a plan moving forward. 

@vsoch I'm not sure how involved you are these days - but would appreciate your thoughts. As well as @johanneskoester and @dlaehnemann.

**Comment by dlaehnemann** [comment]
A new executor sounds like the way to go. Just to be sure that you're aware of it, there's some documentation on how to contribute a new one:

https://snakemake.readthedocs.io/en/latest/project_info/contributing.html#contributing-a-new-cluster-or-cloud-execution-backend

In general, these are best written (and debugged) by people actually using the respective cloud platform / server cluster system. And if I remember correctly, Google Cloud was not offering any free testing capabilities for Open Source projects (although asking again never hurts), further complicating continuous integration of any such executor and pretty much precluding non-users of the Google Cloud (like Johannes and me) and the project's overall CI from testing stuff. But we'll obviously try and provide feedback and reviews where we can.

**Comment by vsoch** [comment]
@cademirch the behind the scenes update is that the Life Sciences API is being deprecated (there was a public email this morning) and it's being replaced with Google Batch. We have underway a collaboration that is going to be adding Google Batch to snakemake, and deprecating the Lifesciences API. We are also adding a new executor plugin framework that will make it much more flexible to develop / use one-off executor plugins.

TLDR: this contract / work is already underway - the statement of work is getting finished up and I should be able to start working on it soon. This will include deprecating Life Sciences (and moving into a custom executor plugin) and creating a new Google Batch executor plugin. Sound good?

**Comment by vsoch** [comment]
And here is the work on the custom executor plugin: https://github.com/snakemake/snakemake/pull/2305. Johannes is on vacation but should be more responsive next month.

**Comment by cademirch** [comment]
@vsoch to the rescue! Thanks, I wasn't aware of the executor plugin, looks great. 

Will close this, feel free to tag me in anything that needs help/testing.

Thanks!

**Comment by vsoch** [comment]
Absolutely! The only delay would be in paperwork, but hopefully we will get started soon! We will definitely meet the deprecation deadline in 2025 (omg how is it almost close to that even?!)

**Comment by hnawar** [comment]
As an update, any new projects can still enable Life Sciences API, so there is no immediate impact. Hopefully the transition to Batch API will be complete soon and with feature parity

**Comment by jjjermiah** [comment]
Hi there! 
I am getting started with some heavy processing and having access to some google resource creditds, started looking into Batch as an executor for snakemake seeing how they have already implemented Nextflow with Google Batch. Thankfully found this thread first, and wanted to ask about what that collaboration looks like and if there will be a branch created to implement this feature that the community can contribute to? 

Thank you!

**Comment by vsoch** [comment]
hey @jjjermiah ! We should be starting soon (I anticipate in a week or two) - there is a lot of paper work to figure out when something like this happens, but I actually think we are almost done! So I (think) I should get the green flag soon. What will happen is that we will create a separate repository for a batch plugin, and would you like me to ping you when there is something ready for testing?

For what you are interested in - if you want to get started immediately you could use the Google Lifesciences Executor, although likely soon I'm going to be moving it into an executor plugin using our new plugin setup for Snakemake. Pinging @johanneskoester for his thoughts.

**Comment by jjjermiah** [comment]
@vsoch please do ping me! its not a rush at all, my workflow currently uses a GKE cluster. I'm still learning the ins and outs of GCP but batch looks very powerful as I'm looking into use cases. 

I've started parsing through the executor plugin. Would be very interested to see how this batch executor is developed from scratch, I'm hoping I can contribute wherever possible!

**Comment by vsoch** [comment]
@jjjermiah I'm hoping it doesn't take me too long - @johanneskoester and I designed the executor plugin setup, and I've written already many Python SDK examples with batch, so I suppose I just need to pancake those two things together :laughing: :pancakes:

**Comment by cademirch** [comment]
@vsoch if you could also tag me in the Batch executor repo when you start it, I would greatly appreciate it!

**Comment by hnawar** [comment]
@vsoch Please let me know as well!

**Comment by cademirch** [comment]
@vsoch Is the batch executor public? I've got some ideas regarding preemptive instances and logging from my groups experience with GLS, and it be awesome to contribute. 

I'll take a look at migrating the Google storage code this week - should that belong in the snakemake GH account? I guess I could make one then transfer ownership later.

**Comment by vsoch** [comment]
Yes and I’ll share it for feedback when I’m ready for that, thanks! We have a very specific spec, detailed by Google, and I don’t want to diverge from that to start.

As I mentioned I’d be happy to make a repository here with the scaffolding.

**Comment by cademirch** [comment]
Got it, that makes sense! And yes if you start a repo that would be great.

**Comment by vsoch** [comment]
Fantastic! I got a good start on it yesterday and will put it online soon - likely in the next few hours at most or sooner.

**Comment by vsoch** [comment]
okay - ready for you! 

https://github.com/snakemake/snakemake-storage-plugin-gs

I did more yesterday than I remembered - but caution - this is all written code with zero testing! If you have questions for how to get it working let me know! Generally you can specify to use "gs" as the default storage after installing your plugin (and make sure you install directly and not in development mode) and then go from there. What I do is insert IPython.embeds where I want to look around and develop. For a first step you might try to make a basic test case that was used previously in the snakemake repository. When you are ready for a review / another set of eyes (or want to discuss) please open issues on the repository there and I can assist. Thank you!

**Comment by vsoch** [comment]
Heads up I'm doing more work on it now - likely will have some changes pushed this evening!

Wasn't planning on it, but pulling large containers == huge opportunity for double tasking :laughing:
